File: web/src/__tests__/dsl/cache/StalenessChecker.test.ts
import { describe, it, expect } from 'vitest'
import { isStale, shouldReExecute } from '../../../dsl/cache/StalenessChecker'
import { cachedResultSuccess, cachedResultError } from '../../../dsl/cache/CachedDirectiveResult'
import { EMPTY_DEPENDENCIES, type DirectiveDependencies } from '../../../dsl/cache/DirectiveDependencies'
import { EMPTY_METADATA_HASHES } from '../../../dsl/cache/MetadataHasher'
import { computePathHash, computeExistenceHash, computeModifiedHash, computeAllNamesHash } from '../../../dsl/cache/MetadataHasher'
import { hashFirstLine, hashNonFirstLine } from '../../../dsl/cache/ContentHasher'
import { numberVal } from '../../../dsl/runtime/DslValue'
import type { Note } from '../../../data/Note'
import { Timestamp } from 'firebase/firestore'

function makeNote(id: string, content = 'Note content', path = `path/${id}`): Note {
  return {
    id,
    userId: 'user1',
    parentNoteId: null,
    content,
    createdAt: Timestamp.fromMillis(1000),
    updatedAt: Timestamp.fromMillis(2000),
    lastAccessedAt: Timestamp.fromMillis(3000),
    tags: [],
    containedNotes: [],
    state: null,
    path,
    rootNoteId: null,
  }
}

describe('StalenessChecker', () => {
  describe('isStale with no dependencies', () => {
    it('is not stale when no dependencies', () => {
      const cached = cachedResultSuccess(numberVal(42), EMPTY_DEPENDENCIES)
      const notes = [makeNote('a')]
      expect(isStale(cached, notes)).toBe(false)
    })
  })

  describe('isStale with metadata dependencies', () => {
    it('detects staleness when existence changes', () => {
      const notes = [makeNote('a'), makeNote('b')]
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, dependsOnNoteExistence: true }
      const metadataHashes = { ...EMPTY_METADATA_HASHES, existenceHash: computeExistenceHash(notes) }
      const cached = cachedResultSuccess(numberVal(1), deps, new Map(), metadataHashes)

      // Same notes - not stale
      expect(isStale(cached, notes)).toBe(false)

      // Add a note - stale
      const moreNotes = [...notes, makeNote('c')]
      expect(isStale(cached, moreNotes)).toBe(true)
    })

    it('detects staleness when path changes', () => {
      const notes = [makeNote('a', 'content', 'old/path')]
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, dependsOnPath: true }
      const metadataHashes = { ...EMPTY_METADATA_HASHES, pathHash: computePathHash(notes) }
      const cached = cachedResultSuccess(numberVal(1), deps, new Map(), metadataHashes)

      const modified = [makeNote('a', 'content', 'new/path')]
      expect(isStale(cached, modified)).toBe(true)
    })

    it('detects staleness when modified timestamp changes', () => {
      const notes = [makeNote('a')]
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, dependsOnModified: true }
      const metadataHashes = { ...EMPTY_METADATA_HASHES, modifiedHash: computeModifiedHash(notes) }
      const cached = cachedResultSuccess(numberVal(1), deps, new Map(), metadataHashes)

      const modified = [{ ...notes[0]!, updatedAt: Timestamp.fromMillis(99999) }]
      expect(isStale(cached, modified)).toBe(true)
    })

    it('detects staleness when all names change', () => {
      const notes = [makeNote('a', 'Original Name\nBody')]
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, dependsOnAllNames: true }
      const metadataHashes = { ...EMPTY_METADATA_HASHES, allNamesHash: computeAllNamesHash(notes) }
      const cached = cachedResultSuccess(numberVal(1), deps, new Map(), metadataHashes)

      const modified = [makeNote('a', 'Changed Name\nBody')]
      expect(isStale(cached, modified)).toBe(true)
    })
  })

  describe('isStale with content dependencies', () => {
    it('detects staleness when first line changes', () => {
      const note = makeNote('a', 'First line\nSecond line')
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, firstLineNotes: new Set(['a']) }
      const contentHashes = new Map([['a', { firstLineHash: hashFirstLine(note.content), nonFirstLineHash: null }]])
      const cached = cachedResultSuccess(numberVal(1), deps, contentHashes)

      // Change first line
      const modified = [makeNote('a', 'Different first\nSecond line')]
      expect(isStale(cached, modified)).toBe(true)
    })

    it('not stale when first line is same', () => {
      const note = makeNote('a', 'Same first\nSecond line')
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, firstLineNotes: new Set(['a']) }
      const contentHashes = new Map([['a', { firstLineHash: hashFirstLine(note.content), nonFirstLineHash: null }]])
      const cached = cachedResultSuccess(numberVal(1), deps, contentHashes)

      // Same first line, different body
      const modified = [makeNote('a', 'Same first\nDifferent body')]
      expect(isStale(cached, modified)).toBe(false)
    })

    it('detects staleness when non-first line changes', () => {
      const note = makeNote('a', 'First line\nSecond line')
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, nonFirstLineNotes: new Set(['a']) }
      const contentHashes = new Map([['a', { firstLineHash: null, nonFirstLineHash: hashNonFirstLine(note.content) }]])
      const cached = cachedResultSuccess(numberVal(1), deps, contentHashes)

      const modified = [makeNote('a', 'First line\nChanged second')]
      expect(isStale(cached, modified)).toBe(true)
    })

    it('detects staleness when tracked note is deleted', () => {
      const note = makeNote('a', 'Content')
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, firstLineNotes: new Set(['a']) }
      const contentHashes = new Map([['a', { firstLineHash: hashFirstLine(note.content), nonFirstLineHash: null }]])
      const cached = cachedResultSuccess(numberVal(1), deps, contentHashes)

      // Note 'a' is gone
      const modified = [makeNote('b', 'Other')]
      expect(isStale(cached, modified)).toBe(true)
    })
  })

  describe('shouldReExecute', () => {
    it('returns false for fresh non-stale result', () => {
      const cached = cachedResultSuccess(numberVal(42), EMPTY_DEPENDENCIES)
      expect(shouldReExecute(cached, [])).toBe(false)
    })

    it('returns true for non-deterministic error', () => {
      const error = { kind: 'NetworkError' as const, message: 'Network error', isDeterministic: false }
      const cached = cachedResultError(error)
      expect(shouldReExecute(cached, [])).toBe(true)
    })

    it('returns false for deterministic error', () => {
      const error = { kind: 'TypeError' as const, message: 'Type error', isDeterministic: true }
      const cached = cachedResultError(error)
      expect(shouldReExecute(cached, [])).toBe(false)
    })

    it('returns true when dependencies are stale', () => {
      const notes = [makeNote('a')]
      const deps: DirectiveDependencies = { ...EMPTY_DEPENDENCIES, dependsOnNoteExistence: true }
      const metadataHashes = { ...EMPTY_METADATA_HASHES, existenceHash: computeExistenceHash(notes) }
      const cached = cachedResultSuccess(numberVal(1), deps, new Map(), metadataHashes)

      const moreNotes = [...notes, makeNote('b')]
      expect(shouldReExecute(cached, moreNotes)).toBe(true)
    })
  })
})

File: app/src/test/java/org/alkaline/taskbrain/dsl/cache/StalenessCheckerTest.kt
package org.alkaline.taskbrain.dsl.cache

import com.google.firebase.Timestamp
import org.alkaline.taskbrain.data.Note
import org.alkaline.taskbrain.dsl.runtime.values.StringVal
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.util.Date

/**
 * Tests for StalenessChecker.
 * Phase 1: Data structures and hashing infrastructure.
 */
class StalenessCheckerTest {

    @Before
    fun setUp() {
        // Phase 3: Clear cached metadata hashes before each test
        MetadataHasher.invalidateCache()
    }

    // region Metadata staleness

    @Test
    fun `not stale when no dependencies`() {
        val cached = createCachedResult(DirectiveDependencies.EMPTY)
        val notes = listOf(createNote("1", "inbox"))

        assertFalse(StalenessChecker.isStale(cached, notes))
    }

    @Test
    fun `stale when note existence changes - note added`() {
        val notes1 = listOf(createNote("1", "inbox"))
        val deps = DirectiveDependencies(dependsOnNoteExistence = true)
        val hashes = MetadataHashes(existenceHash = MetadataHasher.computeExistenceHash(notes1))
        val cached = createCachedResult(deps, metadataHashes = hashes)

        val notes2 = notes1 + createNote("2", "archive")

        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    @Test
    fun `stale when note existence changes - note removed`() {
        val notes1 = listOf(createNote("1", "inbox"), createNote("2", "archive"))
        val deps = DirectiveDependencies(dependsOnNoteExistence = true)
        val hashes = MetadataHashes(existenceHash = MetadataHasher.computeExistenceHash(notes1))
        val cached = createCachedResult(deps, metadataHashes = hashes)

        val notes2 = listOf(createNote("1", "inbox"))

        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    @Test
    fun `stale when path changes`() {
        val notes1 = listOf(createNote("1", "inbox"))
        val deps = DirectiveDependencies(dependsOnPath = true)
        val hashes = MetadataHashes(pathHash = MetadataHasher.computePathHash(notes1))
        val cached = createCachedResult(deps, metadataHashes = hashes)

        val notes2 = listOf(createNote("1", "archive"))

        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    @Test
    fun `not stale when path same`() {
        val notes = listOf(createNote("1", "inbox"))
        val deps = DirectiveDependencies(dependsOnPath = true)
        val hashes = MetadataHashes(pathHash = MetadataHasher.computePathHash(notes))
        val cached = createCachedResult(deps, metadataHashes = hashes)

        assertFalse(StalenessChecker.isStale(cached, notes))
    }

    @Test
    fun `stale when modified timestamp changes`() {
        val time1 = Timestamp(Date(1000L))
        val time2 = Timestamp(Date(2000L))
        val notes1 = listOf(createNote("1", "inbox", updatedAt = time1))
        val deps = DirectiveDependencies(dependsOnModified = true)
        val hashes = MetadataHashes(modifiedHash = MetadataHasher.computeModifiedHash(notes1))
        val cached = createCachedResult(deps, metadataHashes = hashes)

        val notes2 = listOf(createNote("1", "inbox", updatedAt = time2))

        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    // endregion

    // region Content staleness

    @Test
    fun `stale when first line changes`() {
        val note1 = createNote("1", "inbox", content = "Old name\nBody")
        val deps = DirectiveDependencies(firstLineNotes = setOf("1"))
        val contentHashes = mapOf(
            "1" to ContentHashes(firstLineHash = ContentHasher.hashFirstLine(note1.content))
        )
        val cached = createCachedResult(deps, noteContentHashes = contentHashes)

        val note2 = createNote("1", "inbox", content = "New name\nBody")

        assertTrue(StalenessChecker.isStale(cached, listOf(note2)))
    }

    @Test
    fun `not stale when first line same but body changes`() {
        val note1 = createNote("1", "inbox", content = "Same name\nOld body")
        val deps = DirectiveDependencies(firstLineNotes = setOf("1"))
        val contentHashes = mapOf(
            "1" to ContentHashes(firstLineHash = ContentHasher.hashFirstLine(note1.content))
        )
        val cached = createCachedResult(deps, noteContentHashes = contentHashes)

        val note2 = createNote("1", "inbox", content = "Same name\nNew body")

        assertFalse(StalenessChecker.isStale(cached, listOf(note2)))
    }

    @Test
    fun `stale when non-first line changes`() {
        val note1 = createNote("1", "inbox", content = "Name\nOld body")
        val deps = DirectiveDependencies(nonFirstLineNotes = setOf("1"))
        val contentHashes = mapOf(
            "1" to ContentHashes(nonFirstLineHash = ContentHasher.hashNonFirstLine(note1.content))
        )
        val cached = createCachedResult(deps, noteContentHashes = contentHashes)

        val note2 = createNote("1", "inbox", content = "Name\nNew body")

        assertTrue(StalenessChecker.isStale(cached, listOf(note2)))
    }

    @Test
    fun `stale when dependent note is deleted`() {
        val note1 = createNote("1", "inbox", content = "Content")
        val deps = DirectiveDependencies(firstLineNotes = setOf("1"))
        val contentHashes = mapOf(
            "1" to ContentHashes(firstLineHash = ContentHasher.hashFirstLine(note1.content))
        )
        val cached = createCachedResult(deps, noteContentHashes = contentHashes)

        // Note "1" no longer exists
        val notes2 = listOf(createNote("2", "archive"))

        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    // endregion

    // region Short-circuit behavior

    @Test
    fun `short circuits on first stale field`() {
        // This test verifies the short-circuit behavior exists
        // by having multiple stale conditions
        val notes1 = listOf(createNote("1", "path1"))
        val deps = DirectiveDependencies(
            dependsOnPath = true,
            dependsOnNoteExistence = true
        )
        val hashes = MetadataHashes(
            pathHash = MetadataHasher.computePathHash(notes1),
            existenceHash = MetadataHasher.computeExistenceHash(notes1)
        )
        val cached = createCachedResult(deps, metadataHashes = hashes)

        // Both existence (new note) and path (changed) are stale
        val notes2 = listOf(
            createNote("1", "path2"),
            createNote("2", "path3")
        )

        // Should return true without checking all conditions
        assertTrue(StalenessChecker.isStale(cached, notes2))
    }

    // endregion

    private fun createNote(
        id: String,
        path: String,
        content: String = "Content",
        updatedAt: Timestamp? = null
    ) = Note(
        id = id,
        path = path,
        content = content,
        updatedAt = updatedAt
    )

    private fun createCachedResult(
        deps: DirectiveDependencies,
        noteContentHashes: Map<String, ContentHashes> = emptyMap(),
        metadataHashes: MetadataHashes = MetadataHashes.EMPTY
    ) = CachedDirectiveResult(
        result = StringVal("cached"),
        dependencies = deps,
        noteContentHashes = noteContentHashes,
        metadataHashes = metadataHashes
    )
}
